pkg/net/bitswap.go

```diff
@@ -151,3 +163,161 @@ func (c BitswapClient) Retrieve(
 		return task.NewErrorRetrievalResultWithErrorResolution(task.RetrievalFailure, err), nil
 	}
 }
+
+// Starts with the root CID, then fetches a random CID from the children and grandchildren nodes, until it reaches `traverseDepth`
+// Note: the root CID is considered depth `0`, so passing `traverseDepth=0` will only fetch the root CID
+// Returns a `SuccessfulRetrievalResult` if *all* retrievals were successful, `ErrorRetrievalResult` if any failed
+func (c BitswapClient) SpadeTraversal(parent context.Context, target peer.AddrInfo, startingCid cid.Cid, traverseDepth uint) (*task.RetrievalResult, error) {
+	logger := logging.Logger("bitswap_client_spade").With("cid", startingCid).With("target", target)
+	cidToRetrieve := startingCid
+
+	startTime := time.Now()
+
+	// support structures such as: https://github.com/filecoin-project/go-dagaggregator-unixfs#grouping-unixfs-structure
+	i := uint(0)
+	for {
+		// For some reason, need to re-init the host & client every time we do a fetch
+		// otherwise, we get context timeout error after the first fetch
+		host, err := InitHost(parent, nil)
+		if err != nil {
+			return task.NewErrorRetrievalResult(task.CannotConnect, errors.Wrap(err, "failed to init host %s")), nil
+		}
+
+		client := NewBitswapClient(host, time.Second*1)
+
+		// Retrieval
+		logger.Infof("retrieving %s\n", cidToRetrieve.String())
+		blk, err := client.RetrieveBlock(parent, target, cidToRetrieve)
```

Comment:
RetrieveBlock creates network and bsnet objects which ideally should be created once over the whole retrieval. You may need to refactor it to create those objects outside of any for loop.

Reply:
I just did this and it seems to have solved the other issue with the context timeouts 👍